query-engine/core/src/interactive_transactions/manager.rs

```diff
@@ -0,0 +1,192 @@
+
+    /// Sender part of the channel to which transaction id is sent when the timeout of the
+    /// transaction expires.
+    timeout_sender: UnboundedSender<TxId>,
+}
+
+impl ItxManager {
+    pub fn new() -> Self {
+        let transactions = Arc::new(RwLock::new(HashMap::<_, Arc<Mutex<InteractiveTransaction>>>::default()));
+        let closed_txs = Arc::new(RwLock::new(LruCache::new(*CLOSED_TX_CACHE_SIZE)));
+        let (timeout_sender, mut timeout_receiver) = unbounded_channel();
+
+        // This task rollbacks and removes any open transactions with expired timeouts from the
+        // `self.transactions`. It also removes any closed transactions to avoid `self.transactions`
+        // growing infinitely in size over time.
+        // Note that this task automatically exits when all transactions finish and the `ItxManager`
+        // is dropped, because that causes the `timeout_receiver` to become closed.
+        crosstarget_utils::task::spawn({
+            let transactions = Arc::clone(&transactions);
+            let closed_txs = Arc::clone(&closed_txs);
+            async move {
+                while let Some(tx_id) = timeout_receiver.recv().await {
+                    let transaction_entry = match transactions.write().await.remove(&tx_id) {
+                        Some(transaction_entry) => transaction_entry,
+                        None => {
+                            // Transaction was committed or rolled back already.
+                            continue;
+                        }
+                    };
+                    let mut transaction = transaction_entry.lock().await;
+
+                    // If transaction was already committed, rollback will error.
+                    let _ = transaction.rollback(true).await;
+
+                    let closed_tx = transaction
+                        .as_closed()
+                        .expect("transaction must be closed after rollback");
+
+                    closed_txs.write().await.put(tx_id, closed_tx);
+                }
+            }
+            .with_current_subscriber()
+            .with_current_recorder()
+        });
+
+        Self {
+            transactions,
+            closed_txs,
+            timeout_sender,
+        }
+    }
+
+    pub async fn create_tx(
+        &self,
+        query_schema: QuerySchemaRef,
+        tx_id: TxId,
+        conn: Box<dyn Connection + Send + Sync>,
+        isolation_level: Option<String>,
+        timeout: Duration,
+    ) -> crate::Result<()> {
+        // This task notifies the task spawned in `new()` method that the timeout for this
+        // transaction has expired.
+        crosstarget_utils::task::spawn({
+            let timeout_sender = self.timeout_sender.clone();
+            let tx_id = tx_id.clone();
+            async move {
+                crosstarget_utils::time::sleep(timeout).await;
+                timeout_sender.send(tx_id).expect("receiver must exist");
+            }
+        });
```

Comment:
I didn't add `with_subscriber().with_recorder()` here because neither is used in the task and it's fairly small